### ISSUE TITLE
ctypes find_library does not find libspatialindex_c.so in non-standard locations

### ISSUE THREAD

**Issue by samfux84** [Toblerity/rtree#56]
Please do not use find_library from ctypes to find libspatialindex_c. This only works if it is installed on in a standard location. On HPC clusters, it is very likely that the library is not installed in a standard location. Adding the installation directory to LD_LIBRARY_PATH, or specifying the installation directory in the setup.cfg file should be sufficient, but it does not work.

The only way to make rtree find libspatialindex_c on a Linux HPC cluster is to "hack" core.py, comment out the line

lib_name = find_library('spatialindex_c')

and replace lib_name in

rt = ctypes.CDLL(lib_name)

with the absolut path of the library.

I understand that the problem is introduced by ctypes and not by the rtree package, but please use another mechanism than find_library to detect libspatialindex_c on Linux systems. The two most practical options would be to either use LD_LIBRARY_PATH or the setup.cfg file.

Thank you very much and best regards

Samuel Fux

**Comment by hobu** [comment]
What is the Python Way™ to do this nowadays? The ctypes find_library code is from eons ago (taken from Shapely which took it from something else IIRC).

We need something that doesn't interfere with virtualenv and installs via pip to the right location.

**Comment by samfux84** [comment]
In my opinion, it is still ok to use ctypes find_library (it seems to work on all operating systems except Linux), but if find_library returns nothing, then it would be nice if there was a fallback option that opens the setup.cfg file and checks if the path to libspatialindex_c is specified there.

**Comment by hobu** [comment]
Please make a pull request implementing as you propose.

**Comment by mwtoews** [comment]
I can confirm that if spatialindex is build with `./configure --prefix=/some/dir`, and `/some/dir` is on `LD_LIBRARY_PATH`, then `ctypes.CDLL('libspatialindex_c.so')` correctly finds the library, while `find_library` does not work (returns None). Perhaps [these lines in core.py](https://github.com/Toblerity/rtree/blob/6f2e78454b567cc70e5c3d09caa3a5373c508656/rtree/core.py#L103-L109) should be changed:
```diff
@@ -103,5 +103,3 @@
 elif os.name == 'posix':
-    platform = os.uname()[0]
-    lib_name = find_library('spatialindex_c')
-    rt = ctypes.CDLL(lib_name)
+    rt = ctypes.CDLL('libspatialindex_c.so')
 else:
```

**Comment by samfux84** [comment]
I did a bit more research on this topic and changed my mind. It is not rtree that needs to be changed because the installation routines of rtree do the correct thing. Ctypes needs to be fixed as the bug is in there. The bug is known since 2010 (https://bugs.python.org/issue9998) and people even provided patches for it. After more than 5 years, the status of the bug is still "open" and the patches did not make it into the official Python distribution. Therefore it seems that nobody really cares ...

I am not tilting at windmills since I don't have time for that. In the first post above I provided a workaround which solves the problem for people installing rtree on a Linux machine.

**Comment by hobu** [comment]
@mwtoews We can't change to your proposed code because OSX and other *nix sometimes do not end in `.so` for shared libraries (`.dylib` in that case). We'd end up reimplementing find_library, poorly.

I'll happily merge something that does what you need and doesn't break everyone else. A hard balance to maintain...

**Comment by mwtoews** [comment]
@hobu correct, that'd break OS X.

I can prepare a patch over the next week or two, learning again from Shapely's use (e.g. either [libgeos.py](https://github.com/Toblerity/Shapely/blob/2b631d3c7065c700a56e6dc3f5dee18117b93c03/shapely/libgeos.py) or [_buildcfg.py](https://github.com/Toblerity/Shapely/blob/06e89d121cd127d4b262b432df52f855d1125319/shapely/_buildcfg.py)). I think it would also be a good idea to introduce a generic environment variable such as `SPATIALINDEX_LIBRARY_PATH` to point to a shared library (.so, .dll, etc.).

**Comment by hobu** [comment]
As long as `SPATIALINDEX_LIBRARY_PATH` isn't required, that sounds like a good plan.

**Comment by vmora** [comment]
The two env variables in #73 are merged, shall this one be closed ?

**Comment by samfux84** [comment]
Thank you very much for implementing the env variable. I think this is a very good solution for this problem.

**Comment by citibeth** [comment]
I'm new to this rtree, but just ran into this issue as well.  I don't think env vars are a good solution because then the env var has to be set correctly in order for the system to work.  Google "LD_LIBRARY_PATH evil" to see a broader discussion on this issue.

Preferable would be an analog to RPATH; i.e. bake the location of libspatialindex into rtree when you build/install.

**Comment by sgillies** [comment]
Interestingly, cffi has no silver bullets for finding library paths: http://cffi.readthedocs.io/en/latest/cdef.html#ffi-dlopen-loading-libraries-in-abi-mode.

Closing this issue. See #74 for related work.

**Comment by snorfalorpagus** [comment]
The issue reference by @samfux84 has now been "fixed": https://bugs.python.org/issue9998

As of Python 3.6, on Linux the `find_library` function will search `LD_LIBRARY_PATH`. I don't think this changes anything for us immediately, but it's good news for the future.

**Comment by citibeth** [comment]
I agree, it doesn't change anything immediately.  You shouldn't have to have LD_LIBRARY_PATH set in order for this to work.

**Comment by linwoodc3** [comment]
Greetings:

I also received this error today (31 Jan 2017):
```bash
OSError: libspatialindex_c.so: cannot open shared object file: No such file or directory
```
I am on a CentOS 6.7 machine, running Anaconda 4,, Python 2.7.12.  Not the most-up-to-date system but it's what I have :-(.  
**TLDR**:  How do I install rtree from the tar.gz or local whl and set this path.  I did the `./configure ->make->make install` for libspatialindex and had no errors, and saw it in `anaconda/lib` dir.  Need exact install command to run from bash/terminal please :-)
**Details**
This install needs to happen on an airgapped Linux box; 

Why: I am using [`geopandas`](http://geopandas.org/index.html), which has `fiona` and `shapely` as dependencies.  The [`geopandas spatial join`](http://geopandas.org/mergingdata.html#spatial-joins)  requires `rtree`, so I NEED `rtree` or else I do a predicate operation on each point-polygon comparision which is VERY inefficient.  So, **bottom-line**, hoping the smart folks on this issue can tell me which method above works (I'll write it down and use it on my machine). I see LD_LIBRARY_PATH mentioned but, stupid guy proof (me), trying to get the exact command to run.Didn't see it above so having the solution explicitly defined as `run this` would help other folks who find this issue like me. 

**Extra Detail**
I have the Rtree tar.gz downloaded, and confirmed that libspatialindex_c.so is in my `/home/<username>/anaconda2/lib/` directory.

**Comment by samfux84** [comment]
Hi linwoodc3,

download the source code of Rtree, unpack it, open setup.py in a texteditor. Then replace the line

rt = ctypes.CDLL(lib_name)

with

rt = ctypes.CDLL('libspatialindex_c.so')

Then tar and zip the directory again and install the new tar with pip.

**Comment by linwoodc3** [comment]
Worked beautifully.  For any who come after me, with the same problem:

* Downloaded this spatialindex or version you need; I used [1.8.5 version here ](https://libspatialindex.github.io/#download)
* Followed Autotools instructions here: https://libspatialindex.github.io/install.html
* * ran `pip download rtree` .  This gave me **Rtree-0.8.2.tar.gz**.  Untar this file
* Followed [samfux84 instruction comment directly above](https://github.com/Toblerity/rtree/issues/56#issuecomment-276584781)
* Retar file with `tar -czvf Rtree-0.8.2.tar.gz Rtree-0.8.2/`
* `pip install Rtree-0.8.2.tar.gz`

**Comment by frederickayala** [comment]
On Linux, adding the following line to my ~/.bashrc fixed the issue:

export LD_LIBRARY_PATH="/usr/local/lib/:LD_LIBRARY_PATH"

**Comment by palewire** [comment]
I got this error trying to do a spatial join in geopandas, which I believe relies on Rtree. I am inside a virtual environment, which I know can sometimes cause these type of problems.

```python
---------------------------------------------------------------------------
OSError                                   Traceback (most recent call last)
<ipython-input-50-ed85e7de143b> in <module>()
----> 1 ca_approved_gdf_with_county = gp.sjoin(ca_approved_gdf, counties, how="inner", op='intersects')

/home/ben/.virtualenvs/h2-visas-analysis/local/lib/python2.7/site-packages/geopandas/tools/sjoin.pyc in sjoin(left_df, right_df, how, op, lsuffix, rsuffix)
     27 
     28     """
---> 29     import rtree
     30 
     31     allowed_hows = ['left', 'right', 'inner']

/home/ben/.virtualenvs/h2-visas-analysis/local/lib/python2.7/site-packages/rtree/__init__.py in <module>()
----> 1 from .index import Rtree
      2 
      3 from .core import rt
      4 
      5 __version__ = '0.8.3'

/home/ben/.virtualenvs/h2-visas-analysis/local/lib/python2.7/site-packages/rtree/index.py in <module>()
      3 import pprint
      4 
----> 5 from . import core
      6 import ctypes
      7 try:

/home/ben/.virtualenvs/h2-visas-analysis/local/lib/python2.7/site-packages/rtree/core.py in <module>()
    123 
    124     if lib_name is None:
--> 125         raise OSError("Could not find libspatialindex_c library file")
    126 
    127     rt = ctypes.CDLL(lib_name)

OSError: Could not find libspatialindex_c library file
```

**Comment by amarjeetkapoor1** [comment]
I am getting this error on installing with pip and easy install.

**Comment by linwoodc3** [comment]
I'm sorry if this is a silly question @palewire , but [did you try to do the steps above?](https://github.com/Toblerity/rtree/issues/56#issuecomment-276584781)

**Comment by jczaplew** [comment]
@frederickayala's solution worked for me. Just don't forget to `source ~/.bashrc` as well

**Comment by rishabh-90** [comment]
@palewire did you get it work I am also getting same error

**Comment by rishabh-90** [comment]
@palewire ok I found answer to this and it is working trying doing these steps

sudo apt-get update
sudo apt-get install curl
sudo apt-get install g++
sudo apt-get install make

curl -L http://download.osgeo.org/libspatialindex/spatialindex-src-1.8.5.tar.gz | tar xz
cd spatialindex-src-1.8.5
./configure
make
sudo make install
sudo ldconfig

**Comment by rmania** [comment]
/.configure file won't be there if you don' t first run ./autogen.sh so this step needed to be induded in my case;

sudo apt-get update
sudo apt-get install curl
sudo apt-get install g++
sudo apt-get install make

curl -L http://download.osgeo.org/libspatialindex/spatialindex-src-1.8.5.tar.gz | tar xz
cd spatialindex-src-1.8.5
./autogen.sh
./configure
make
sudo make install
sudo ldconfig
